src/palimpzest/solver/query_strategies.py

```diff
@@ -0,0 +1,258 @@
+            # parse the json object and set the DataRecord's fields with their generated values 
+            setattr(dr, field_name, jsonObj[field_name])
+
+    return dr
+
+# TODO: Stats
+def runBondedQuery(candidate: DataRecord, td: TaskDescriptor, verbose: bool=False) -> Tuple[List[DataRecord], dict, str]:
+    """
+    Run a bonded query, in which all new fields in the outputSchema are generated simultaneously
+    in a single LLM call. This is in contrast to a conventional query, in which each output field
+    is generated using its own LLM call.
+
+    At the moment, tasks with cardinality == "oneToMany" can only be executed using bonded queries.
+
+    This is not a theoretical limitation of conventional queries, but there are some practical
+    difficulties with guaranteeing that each field has the same number of outputs generated
+    in each separate LLM invocation.
+    """
+    # initialize list of output data records
+    drs = []
+
+    # construct list of fields in outputSchema which will need to be generated
+    generate_field_names = []
+    for field_name in td.outputSchema.fieldNames():
+        if field_name not in td.inputSchema.fieldNames():
+            generate_field_names.append(field_name)
+
+    # fetch input information
+    text_content = candidate.asTextJSON()
+    doc_schema = str(td.outputSchema)
+    doc_type = td.outputSchema.className()
+
+    # construct prompt question
+    promptQuestion = _construct_query_prompt(td, doc_type, generate_field_names)
+
+    # generate LLM response and capture statistics
+    answer, query_stats = None, None
+    try:
+        if td.prompt_strategy == PromptStrategy.DSPY_COT_QA:
+            # invoke LLM to generate output JSON
+            generator = DSPyGenerator(td.model.value, td.prompt_strategy, doc_schema, doc_type, verbose)
+            answer, query_stats = generator.generate(text_content, promptQuestion)
+
+            # add input and *computed* output fields to query_stats object
+            query_stats["in_fields"] = td.inputSchema.fieldNames()
+            query_stats["out_fields"] = generate_field_names
+
+        elif td.prompt_strategy == PromptStrategy.IMAGE_TO_TEXT:
+            # b64 decode of candidate.contents
+            image_b64 = base64.b64encode(candidate.contents).decode('utf-8')
+
+            # invoke LLM to generate output JSON
+            generator = ImageTextGenerator(td.model.value)
+            answer, query_stats = generator.generate(image_b64, promptQuestion)
+
+            # add input and *computed* output fields to query_stats object
+            query_stats["in_fields"] = td.inputSchema.fieldNames()
+            query_stats["out_fields"] = generate_field_names
+
+        # TODO
+        elif td.prompt_strategy == PromptStrategy.ZERO_SHOT:
+            raise Exception("not implemented yet")
+
+        # TODO
+        elif td.prompt_strategy == PromptStrategy.FEW_SHOT:
+            raise Exception("not implemented yet")
+
```

Comment:
Where would Chunwei add his optimization that figures out which parts of the input are most important for a particular output?

Reply:
Good question, and it may be something I can answer better after finishing PR2, since some of that will depend on how the re-optimization code works.